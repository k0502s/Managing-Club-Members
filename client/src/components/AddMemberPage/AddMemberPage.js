import React, { useState, useEffect } from 'react';
import { Button, Form, Input } from 'reactstrap';
import FileUpload from '../utils/FileUpload';
import { useDispatch, useSelector } from 'react-redux';
import { MEMBER_UPLOADING_REQUEST } from '../../redux/types';

const Continents = [
    { key: 1, value: '남' },
    { key: 2, value: '여' },
];

const AddMemberPage = (props) => {
    const [Title, setTitle] = useState('');
    const [Description, setDescription] = useState('');
    const [Price, setPrice] = useState('');
    const [Continent, setContinent] = useState(0);
    const [Images, setImages] = useState([]);
    const dispatch = useDispatch();

    const { user } = useSelector((state) => state.auth);

    const titleChangeHandler = (e) => {
        setTitle(e.currentTarget.value);
    };

    const descriptionChangeHandler = (e) => {
        setDescription(e.currentTarget.value);
    };

    const priceChangeHandler = (e) => {
        setPrice(e.currentTarget.value);
    };

    const continentChangeHandler = (e) => {
        setContinent(e.currentTarget.value);
    };

    const updateImages = (newImages) => {
        setImages(newImages);
    };

    const submitHandler = (e) => {
        e.preventDefault();

        const name = document.myform.title.value;
        const camera = document.myform.description.value;
        const age = document.myform.price.value;
        const sex = document.myform.sex.value;
        // if(name==="" || camera==="" || age==="" || sex===""){
        //     return alert("빈 칸에 정보를 입력해야 합니다.")
        // }
        if (name === '') {
            return alert('이름 정보를 입력해야 합니다.');
        }
        if (camera === '') {
            return alert('카메라 기종 정보를 입력해야 합니다.');
        }
        if (age === '') {
            return alert('나이 정보를 입력해야 합니다.');
        }
        if (sex === '') {
            return alert('성별 정보를 입력해야 합니다.');
        }
        if (Images === []) {
            return alert('프로필 사진을 입력해야 합니다.');
        }

        //서버에 채운 값들을 request을 보낸다.
        const body = {
            writer: user._id, //로그인된 사람의 ID
            title: Title,
            description: Description,
            price: Price,
            images: Images,
            continents: Continent,
        };
        dispatch({
            type: MEMBER_UPLOADING_REQUEST,
            payload: body,
        });
    };

    return (
        <div style={{ maxWidth: '700px', margin: '2rem auto' }}>
            <div style={{ textAlign: 'center', marginBottom: '1rem' }}>
                <h2>회원 추가하기</h2>
            </div>
            <Form onSubmit={submitHandler} name="myform">
                {/* DropZone   */}
                <FileUpload refreshFunction={updateImages} />
                <label>이름</label>
                <Input onChange={titleChangeHandler} value={Title} name="title" />
                <br />
                <label>카메라 기종</label>
                <Input onChange={descriptionChangeHandler} value={Description} name="description" />
                <br />
                <label>나이</label>
                <Input type="number" onChange={priceChangeHandler} value={Price} name="price" />
                <br />
                <select onChange={continentChangeHandler} value={Continent} name="sex">
                    <option value="">성별을 선택해주세요</option>
                    {Continents.map((item) => (
                        <option key={item.key} value={item.key}>
                            {item.value}
                        </option>
                    ))}
                </select>
                <Button className="col-md-2 offset-md-10" onClick={submitHandler}>
                    확인
                </Button>
            </Form>
        </div>
    );
};

export default AddMemberPage;
