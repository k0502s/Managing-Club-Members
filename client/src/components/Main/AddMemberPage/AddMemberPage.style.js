import styled from 'styled-components';
import { Row, Card, Button } from 'reactstrap';
import { BsPersonFill, BsUpload, BsTrash } from 'react-icons/bs';

// Icon CSS

const PersonIcon = styled(BsPersonFill)`
    width: 300px;
    height: 240px;
    color: lightgrey;
`;
const UploadIcon = styled(BsUpload)`
    color: white;
    margin-bottom: 5px;
    margin-right: 5px;
`;
const DeleteIcon = styled(BsTrash)`
    color: white;
    margin-bottom: 5px;
    margin-right: 5px;
`;

// Add page CSS

const Addbtn = styled(Button)`
    margin-right: 10px;
    font-weight: bold;
    color: white;
    background-color: ${(props) => props.color};

    &:hover {
        color: white;
        background-color: lightslategray;
    }
`;
const Profile = styled.div`
    width: 300px;
    height: 240px;
    border-radius: 55%;
    border: 1px solid lightgray;
    margin-bottom: 15px;
    display: block;
    margin: 0px auto;
    text-align: center;
    @media only screen and (max-width: 360px) {
        margin-left: -10px;
    }
`;

const Img = styled.img`
    width: 300px;
    height: 240px;
    border-radius: 55%;
    border: 1px solid lightgray;
`;

const FileUpload = styled.div`
    margin-top: 30px;
    margin-bottom: 30px;
`;

const PngImg = styled.img`
    width: 400px;
    height: 432px;
    display: block;
    margin: 0px auto;
    @media only screen and (max-width: 767px) {
        width: 300px;
        height: 332px;
    }
`;

const AddCard = styled(Card)`
    height: 790px;
    background-color: #f4fcfb;
    border-radius: 10px;
    @media only screen and (max-width: 767px) {
        margin-top: 10px;
    }
    @media only screen and (min-width: 768px) and (max-width: 1023px) {
        margin-top: 10px;
    }
`;
const Title = styled(Row)`
    text-align: center;
    display: block;
    margin: -10px 0 18px 0;
    color: #f9e81c;
`;
export { Profile, Img, FileUpload, PngImg, AddCard, DeleteIcon, PersonIcon, UploadIcon, Addbtn, Title };
