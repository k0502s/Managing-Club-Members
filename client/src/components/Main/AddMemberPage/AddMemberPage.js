import React, { useState, useEffect } from 'react';
import { Helmet } from 'react-helmet';
import { Form, Input, CardTitle, CardText, Row, Col, CardHeader, CardBody } from 'reactstrap';
import FileUpload from '../../../utils/FileUpload';
import { Loader } from '../../Loader/Loader';
import { useDispatch, useSelector } from 'react-redux';
import img from '../../../assets/img/imgbin_photography-logo-photographer-png.png';
import { MEMBER_UPLOADING_REQUEST, CLEAR_ERROR_REQUEST, CLEAR_ERROR_REQUEST_1 } from '../../../redux/types';
import * as S from './AddMemberPage.style';

const Sex = [
    { key: 1, value: '남' },
    { key: 2, value: '여' },
];

const AddMemberPage = () => {
    const dispatch = useDispatch();
    const [form, setValues] = useState({
        name: '',
        camera: '',
        age: '',
        sex: '',
    });

    const [Images, setImages] = useState([]);
    const { user } = useSelector((state) => state.auth);
    const { isLoading } = useSelector((state) => state.member);

    useEffect(() => {
        dispatch({
            type: CLEAR_ERROR_REQUEST,
        });
        dispatch({
            type: CLEAR_ERROR_REQUEST_1,
        });
    }, []);

    const onChange = (e) => {
        setValues({
            ...form,
            [e.target.name]: e.target.value,
        });
    };

    const submitHandler = (e) => {
        e.preventDefault();
        const nameinput = document.myform.name.value;
        const camerainput = document.myform.camera.value;
        const ageinput = document.myform.age.value;
        const sexinput = document.myform.sex.value;
        if (nameinput === '') {
            return alert('이름 정보를 입력해야 합니다.');
        }
        if (camerainput === '') {
            return alert('카메라 기종 정보를 입력해야 합니다.');
        }
        if (ageinput === '') {
            return alert('나이 정보를 입력해야 합니다.');
        }
        if (sexinput === '') {
            return alert('성별 정보를 입력해야 합니다.');
        }
        if (Images.length === 0) {
            return alert('프로필 사진을 입력해야 합니다.');
        }

        const { name, camera, age, sex } = form;

        const body = {
            writer: user._id,
            name: name,
            camera: camera,
            age: age,
            sex: sex,
            images: Images,
        };

        dispatch({
            type: MEMBER_UPLOADING_REQUEST,
            payload: body,
        });
    };
    const updateImages = (newImages) => {
        setImages([...Images, newImages]);
    };

    const removefile = () => {
        setImages([]);
    };

    const Body = (
        <>
            <S.Title>
                <h1>ADD MEMBER</h1>
            </S.Title>
            <Row>
                <Helmet title={`회원 추가`} />
                <Col>
                    <S.AddCard>
                        <CardHeader>
                            <strong>참고 사항</strong>
                        </CardHeader>
                        <CardBody>
                            <CardTitle tag="h5">※ 회원 추가 안내</CardTitle>
                            <br />
                            <br />
                            <CardText>1. 앞으로 동호회에서 활동 예정인 동호회 회원분들의 정보를 추가해주세요.</CardText>
                            <br />
                            <CardText>2. 추가하실 때 회원 분의 이름, 성별, 나이, 보유 카메라, 프로필 사진을 꼭 추가해주세요.</CardText>
                            <br />
                            <CardText>3. 회원분이 동호회에서 탈퇴하거나 경고 누적으로 제명이 되면 회원 리스트 페이지에서 회원 정보를 삭제할 수 있습니다.</CardText>
                            <S.PngImg src={img} />
                        </CardBody>
                    </S.AddCard>
                </Col>
                <Col>
                    <S.AddCard>
                        <CardHeader>
                            <strong>회원 정보 추가</strong>
                        </CardHeader>
                        <CardBody>
                            <S.Profile>
                                {Images.map((image, index) => (
                                    <S.Img key={index} src={`${image}`} />
                                ))}
                                {Images.length === 0 ? <S.PersonIcon /> : ''}
                            </S.Profile>
                            <S.FileUpload>
                                <FileUpload refreshFunction={updateImages} removefile={removefile} />
                            </S.FileUpload>
                            <Form onSubmit={submitHandler} name="myform" data-testid="form">
                                <label>이름</label>
                                <Input type="text" onChange={onChange} placeholder={'이름 정보를 입력해주세요.'} value={form.name} name="name" data-testid="add-name" />
                                <br />
                                <label>카메라 기종</label>
                                <Input type="text" onChange={onChange} placeholder={'카메라 정보를 입력해주세요.'} value={form.camera} name="camera" data-testid="add-camera" />
                                <br />
                                <label>나이</label>
                                <Input type="number" onChange={onChange} placeholder={'나이 정보를 입력해주세요.'} value={form.age} name="age" data-testid="add-age" />
                                <br />
                                <select onChange={onChange} value={form.sex} name="sex" data-testid="add-sex">
                                    <option value="">성별을 선택해주세요</option>
                                    {Sex.map((item) => (
                                        <option key={item.key} value={item.key} data-testid="select-option">
                                            {item.value}
                                        </option>
                                    ))}
                                </select>
                                <S.Addbtn color={'#54C5A0'} type="submit" className="col-md-2 offset-md-10 mt-1" onClick={submitHandler} data-testid="add-submit">
                                    확인
                                </S.Addbtn>
                            </Form>
                        </CardBody>
                    </S.AddCard>
                </Col>
            </Row>
        </>
    );

    return <>{isLoading ? Loader : Body}</>;
};

export default AddMemberPage;
