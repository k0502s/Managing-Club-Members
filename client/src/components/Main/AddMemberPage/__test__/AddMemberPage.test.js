import React from 'react';
import AddMemberPage from '../AddMemberPage';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

describe('<AddMemberPage />', () => {
    const initialState = { output: 10 };
    const mockStore = configureStore();
    let store;

    it('matches snapshot', () => {
        store = mockStore(initialState);
        const utils = render(
            <Provider store={store}>
                <AddMemberPage />
            </Provider>
        );
        expect(utils.container).toMatchSnapshot();
    });

    it('matches snapshot', () => {
        store = mockStore(initialState);
        const utils = render(
            <Provider store={store}>
                <AddMemberPage />
            </Provider>
        );
        expect(utils.container).toMatchSnapshot();
    });
});
