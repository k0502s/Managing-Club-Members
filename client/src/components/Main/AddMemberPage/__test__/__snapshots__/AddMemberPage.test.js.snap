// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMemberPage /> matches snapshot 1`] = `
<div>
  <div
    style="max-width: 700px; margin: 2rem auto;"
  >
    <div
      style="text-align: center; margin-bottom: 1rem;"
    >
      <h2>
        회원 추가
      </h2>
    </div>
    <section>
      <button
        class="mr-2 btn btn-secondary"
        tabindex="0"
        type="button"
      >
        <input
          autocomplete="off"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        프로필 업로드
      </button>
      <button
        class="btn-danger btn btn-secondary"
        type="button"
      >
        프로필 제거
      </button>
    </section>
    <div
      class="offset-md-4"
      style="width: 300px; height: 240px; border-radius: 55%; border: 1px solid lightgray;"
    />
    <form
      class=""
      name="myform"
    >
      <label>
        이름
      </label>
      <input
        class="form-control"
        name="name"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        카메라 기종
      </label>
      <input
        class="form-control"
        name="camera"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        나이
      </label>
      <input
        class="form-control"
        name="age"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="number"
        value=""
      />
      <br />
      <select
        name="sex"
      >
        <option
          value=""
        >
          성별을 선택해주세요
        </option>
        <option
          value="1"
        >
          남
        </option>
        <option
          value="2"
        >
          여
        </option>
      </select>
      <button
        class="col-md-2 offset-md-10 btn btn-secondary"
        type="button"
      >
        확인
      </button>
    </form>
  </div>
</div>
`;

exports[`<AddMemberPage /> matches snapshot 2`] = `
<div>
  <div
    style="max-width: 700px; margin: 2rem auto;"
  >
    <div
      style="text-align: center; margin-bottom: 1rem;"
    >
      <h2>
        회원 추가
      </h2>
    </div>
    <section>
      <button
        class="mr-2 btn btn-secondary"
        tabindex="0"
        type="button"
      >
        <input
          autocomplete="off"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        프로필 업로드
      </button>
      <button
        class="btn-danger btn btn-secondary"
        type="button"
      >
        프로필 제거
      </button>
    </section>
    <div
      class="offset-md-4"
      style="width: 300px; height: 240px; border-radius: 55%; border: 1px solid lightgray;"
    />
    <form
      class=""
      name="myform"
    >
      <label>
        이름
      </label>
      <input
        class="form-control"
        name="name"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        카메라 기종
      </label>
      <input
        class="form-control"
        name="camera"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        나이
      </label>
      <input
        class="form-control"
        name="age"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="number"
        value=""
      />
      <br />
      <select
        name="sex"
      >
        <option
          value=""
        >
          성별을 선택해주세요
        </option>
        <option
          value="1"
        >
          남
        </option>
        <option
          value="2"
        >
          여
        </option>
      </select>
      <button
        class="col-md-2 offset-md-10 btn btn-secondary"
        type="button"
      >
        확인
      </button>
    </form>
  </div>
</div>
`;
